lib/api/multiObjectDelete.js

```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                            return callback(null, hasGovernanceBypass, objMD, versionId);
+                        });
+                    }
+                    return callback(null, hasGovernanceBypass, objMD, versionId);
+                },
+                (hasGovernanceBypass, objMD, versionId, callback) => {
+                    // AWS only returns an object lock error if a version id
+                    // is specified, else continue to create a delete marker
+                    if (!versionId || !bucket.isObjectLockEnabled()) {
+                        return callback(null, objMD, versionId);
                     }
-                    if (versionId && objMD.location &&
-                        Array.isArray(objMD.location) && objMD.location[0]) {
-                        // we need this information for data deletes to AWS
-                        // eslint-disable-next-line no-param-reassign
-                        objMD.location[0].deleteVersion = true;
+                    const objLockInfo = new ObjectLockInfo({
+                        mode: objMD.retentionMode,
+                        date: objMD.retentionDate,
+                        legalHold: objMD.legalHold || false,
+                    });
+
+                    // If the object can not be deleted raise an error
+                    if (!objLockInfo.canModifyObject(hasGovernanceBypass)) {
+                        log.debug('trying to delete locked object');
+                        return callback(objectLockedError);
                     }
+
                     return callback(null, objMD, versionId);
-                }),
-            (objMD, versionId, callback) => {
-                // AWS only returns an object lock error if a version id
-                // is specified, else continue to create a delete marker
-                if (!versionId || !bucket.isObjectLockEnabled()) {
-                    return callback(null, null, objMD, versionId);
-                }
-                const hasGovernanceBypass = hasGovernanceBypassHeader(request.headers);
-                if (hasGovernanceBypass && authInfo.isRequesterAnIAMUser()) {
-                    return checkUserGovernanceBypass(request, authInfo, bucket, entry.key, log, error => {
-                        if (error && error.is.AccessDenied) {
-                            log.debug('user does not have BypassGovernanceRetention and object is locked', { error });
-                            return callback(objectLockedError);
+                },
+                (objMD, versionId, callback) => {
+                    const options = preprocessingVersioningDelete(
+                        bucketName, bucket, objMD, versionId, config.nullVersionCompatMode);
+                    const deleteInfo = {};
+                    if (options && options.deleteData) {
+                        deleteInfo.deleted = true;
+                        if ((bucket.getLifecycleConfiguration && !bucket.getLifecycleConfiguration())
```

Comment:
in case lifecycle, I think this would be needed only for _lifecycle transition to cold_ is used, and not even at the moment (we did not implement the delete event there event...) : not sure if we can/want to be more precise, or even just skip the test for now.

@nicolas2bert can you confirm if there are conditions where we need the full metadata in the oplog, for lifecycle?

Reply:
in case lifecycle, I think this would be needed only for _lifecycle transition to cold_ is used, and not even at the moment (we did not implement the delete event there event...) : not sure if we can/want to be more precise, or even just skip the test for now.

@nicolas2bert can you confirm if there are conditions where we need the full metadata in the oplog, for lifecycle?